assembly/src/assembler/mod.rs

```diff
@@ -353,17 +361,32 @@ impl Assembler {
     /// to configure the assembler.
     #[instrument("assemble_with_opts_in_context", skip_all)]
     pub fn assemble_with_options_in_context(
-        &mut self,
+        self,
         source: impl Compile,
         options: CompileOptions,
         context: &mut AssemblyContext,
-    ) -> Result<Program, Report> {
+    ) -> Result<MastForest, Report> {
+        self.assemble_with_options_in_context_impl(source, options, context)
+    }
+
+    /// Implementation of [`Self::assemble_with_options_in_context`] which doesn't consume `self`.
+    ///
+    /// The main purpose of this separation is to enable some tests to access the assembler state
+    /// after assembly.
+    fn assemble_with_options_in_context_impl(
```

Comment:
I'm not sure I follow completely why we need to make other methods consume the assembler (rather than take it as a mutable reference). In my mind, the assembler is bound to a kernel on instantiation, and therefore, we never should need to remove the `MastForest` of the kernel from the assembler (as we do on line 388 below).

The way I'm imagining this could work is that when we need to assemble a program, we just clone `MastForest` of the kernel and then build the assembled program on top of it.

But since this is likely to change in the next PR anyway (i.e., we won't need to track the Kernel's MAST forest in the assembler), we probably don't need to do anything about it now.

Reply:
> The way I'm imagining this could work is that when we need to assemble a program, we just clone MastForest of the kernel and then build the assembled program on top of it.

But then the last (or only) program compiled with an `Assembler` would incur a potentially pretty expensive `clone()`. So I would still expose this method, and a new method that would clone the kernel's `MastForest` (as suggested). But let's see how this plays out in the next PR.

Reply:
Good point! Not sure how expensive this cloning would be - but also, no need to check as I think we won't need to do it at all after the next PR.